Fix typos and grammar in the OTLP spec

moonbox3 opened the pull request:
Fixes #3120 

## Changes

This PR fixes typos and grammatical errors in the OTLP spec. Additionally, there were several places where text contained passive voice. These were updated to active voice to make them easier to follow.

moonbox3 commented:
Does this change require a changelog?

tigrannajaryan commented on specification/protocol/otlp.md:
```diff
@@ -648,50 +652,51 @@ client-side queue).
 ## Future Versions and Interoperability
 
 OTLP will evolve and change over time. Future versions of OTLP must be designed
-and implemented in a way that ensures that clients and servers that implement
```
What's wrong with "in a way that"?

moonbox3 replied:
"In a way that ensures" felt too wordy so I wanted to rewrite it for clarity. I really want it to be "... OTLP must be designed and implemented to ensure that clients..." Do you agree?

moonbox3 replied:
I will just revert it...

tigrannajaryan reviewed (changes requested):
Thank you for the fixes @moonbox3

Some of the changes DO change the meaning, so I am temporarily blocking to fix those.

moonbox3 commented on specification/protocol/otlp.md:
```diff
@@ -97,27 +97,27 @@ to receive a response to each request:
-After sending the request the client MAY wait until the response is received
-from the server. In that case there will be at most only one request in flight
-that is not yet acknowledged by the server.
+After sending the request, the client MAY wait until the server receives a response. 
```
Will fix.

moonbox3 commented on specification/protocol/otlp.md:
```diff
@@ -97,27 +97,27 @@ to receive a response to each request:
-desirable and when the client and the server are connected via very low-latency
-network, such as for example when the client is an instrumented application and
+desirable, and when the client and the server are connected via very low-latency
+network, such as when the client is an instrumented application and
```
"Such as" and "for example" are conveying the same thing.

moonbox3 replied:
You don't agree that "such as for example when the client is an instrumented application..." can be reduced to read better as: "such as when the client is an instrumented application..." ?

moonbox3 commented on specification/protocol/otlp.md:
```diff
@@ -97,27 +97,27 @@ to receive a response to each request:
-from the server. In that case there will be at most only one request in flight
-that is not yet acknowledged by the server.
+After sending the request, the client MAY wait until the server receives a response. 
+In that case, there will be, at most only one request in flight that the server does 
+not yet acknowledge.
```
I'm trying to get away from passive voice, which the original was written in.

moonbox3 replied:
Reverted to the original.

moonbox3 commented:
> Thank you for the fixes @moonbox3
> 
> Some of the changes DO change the meaning, so I am temporarily blocking to fix those.

Thank you for your review, @tigrannajaryan. I appreciate it. I have addressed your comments/suggestions.

moonbox3 commented:
> Btw we don't use comma after `e.g.` nor `i.e.` in general - according to [Wiktionary](https://en.wiktionary.org/wiki/e.g.):
> 
> > "E.g." and its examples are typically set off from the rest of the sentence by punctuation. In US English, "e.g." is usually followed by a comma.[[2]](https://en.wiktionary.org/wiki/e.g.#cite_note-2) It is not followed by a comma in other English-speaking countries.
> 
> So we should probably stick to the comma-less form for now, given we already use that and, _technically_, it's not incorrect.

@carlosalberto Thank you for your comment. Is there a style guide I should be following?

MrAlias commented on specification/protocol/otlp.md:
```diff
@@ -235,17 +235,17 @@ containing 0 value of RetryDelay. Here is a sample Go code to illustrate:
 [BadRequest](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L119).
-Other gRPC status code may be used if it is more appropriate. Here is a sample
-Go code to illustrate:
+If more appropriate, another gRPC status code MAY be used. Here is a 
```
This changes "may" into a key word and changes semantic meaning. Is that intentional?

moonbox3 replied:
Didn't mean for may to be capitalized. Fixing.

MrAlias commented on specification/protocol/otlp.md:
```diff
@@ -189,31 +189,31 @@ the number of spans/data points/log records it rejected.
-- Retryable errors indicate that processing of telemetry data failed and the
-  client SHOULD record the error and may retry exporting the same data. This can
+- Retryable errors indicate that telemetry data processing failed, and the
+  client SHOULD record the error and may retry exporting the same data. For example, this can
```
This should follow the same wrapping length as the rest of the paragraph.

MrAlias commented on specification/protocol/otlp.md:
```diff
@@ -189,31 +189,31 @@ the number of spans/data points/log records it rejected.
-  be dropped. This can happen, for example, when the request contains bad data
-  and cannot be deserialized or otherwise processed by the server. The client
+- Not-retryable errors indicate that telemetry data processing failed, and the
+  client MUST NOT retry sending the same telemetry data. The client MUST drop the telemetry data. 
```
This should follow the same wrapping length as the rest of the paragraph.

MrAlias commented on specification/protocol/otlp.md:
```diff
@@ -500,7 +504,7 @@ in case of a successful response.
 
 If the request is only partially accepted
-(i.e. when the server accepts only parts of the data and rejects the rest), the
+(i.e., when the server accepts only parts of the data and rejects the rest), the
```
```suggestion
(i.e. when the server accepts only parts of the data and rejects the rest), the
```

MrAlias commented on specification/protocol/otlp.md:
```diff
@@ -575,29 +580,29 @@ response.
 
 All other HTTP responses that are not explicitly listed in this document should
-be treated according to HTTP specification.
+be treated according to HTTP specifications.
```
Are there multiple HTTP specifications?

moonbox3 replied:
Sure, there exists more than one "specification." As an example, see: https://www.w3.org/Protocols/Specs.html

MrAlias commented on specification/protocol/otlp.md:
```diff
@@ -625,15 +630,14 @@ The default network port for OTLP/HTTP is 4318.
-In such situation the client SHOULD implement queuing, acknowledgement
-handling and retrying logic per destination. This ensures that servers do not
-block each other. The queues SHOULD reference shared, immutable data to be sent,
+In such a situation, the client SHOULD implement queuing, acknowledgment-handling 
```
Adding the leading comma here makes the second sentence fragment appear as a contextual element. Which does not seem to be the original intent of the specification.

moonbox3 replied:
I don't agree that the comma makes the second sentence fragment appear as a contextual element. "In such situation" is an example of an introductory phrase or clause, which sets the stage for a sentence. A writer may use discretion to determine if the comma is needed.

MrAlias replied:
I think both can be true.

Is there a way to restructure this sentence to remove confusion and clarify the meaning? I'm still not 100% sure if this is a list of things that needs to be addressed "per destination" or if only the retry logic is "per destination".

moonbox3 replied:
I think that's a valid callout. This text is under the heading "Multi-Destination Exporting" so I understand it as:

"The client SHOULD implement (per destination): queuing, acknowledgement-handling, and retry logic."

moonbox3 commented on specification/protocol/otlp.md:
```diff
@@ -97,11 +97,11 @@ to receive a response to each request:
 between the client and the server. Many telemetry collection systems have
 intermediary nodes that the data must travel across until reaching the final
-destination (e.g. application -> agent -> collector -> backend). End-to-end
+destination (e.g., application -> agent -> collector -> backend). End-to-end
```
It seems it would be a good idea to add the "no comma needed after e.g." to some style guide in this project.

moonbox3 commented:
@tigrannajaryan or @MrAlias can you please help review the final changes? Thank you.

moonbox3 commented:
I would prefer for this PR to not be closed due to lack of activity. Can I get some help on it, please? Thank you.

carlosalberto commented:
Thanks @moonbox3 - looks good to go IMHO. 

cc @tigrannajaryan

tigrannajaryan commented on specification/protocol/otlp.md:
````diff
@@ -328,30 +334,30 @@ Here is a sample Go code to illustrate:
 ```
 
-The value of `retry_delay` is determined by the server and is implementation
+The server determines the value of the `retry_delay` and is implementation
 dependant. The server SHOULD choose a `retry_delay` value that is big enough to
````
>The server determines the value of the `retry_delay` and is implementation
>dependant.

Unclear what the `is` in the second part of the sentence refers to (the server? the value?).

moonbox3 replied:
I will revert to the original if it is confusing. I understand the "is" is referring to the value of the retry delay.

tigrannajaryan commented on specification/protocol/otlp.md:
```diff
@@ -625,15 +631,14 @@ The default network port for OTLP/HTTP is 4318.
-In such situation the client SHOULD implement queuing, acknowledgement
-handling and retrying logic per destination. This ensures that servers do not
-block each other. The queues SHOULD reference shared, immutable data to be sent,
+In such a situation, the client SHOULD implement (per destination): queuing,  
+acknowledgment-handling, and retry logic. This ensures that servers do not
```
>In such a situation, the client SHOULD implement (per destination): queuing,  
>acknowledgment-handling, and retry logic.

I think this changed the meaning of what we want to say. The emphasis here is that all these capabilities must be implemented per destination, not that these capabilities must be implemented. Putting "per-destination" in parenthesis no longer conveys that important message.

moonbox3 replied:
I reverted to the original.

tigrannajaryan reviewed (commented):
Just 2 small comments left, the rest looks good to me.

moonbox3 commented:
Thank you all for the help. @tigrannajaryan I addressed your final two comments. Thank you.

tigrannajaryan commented:
Thank you @moonbox3 for making the spec clearer.
